### PR TITLE
COMMON: Allow meta engines to modify gui options

### PR DESCRIPTION
**Pull request by athrxx**
This allows adding/removing gui options on the meta engine level.
After parsing the options from the config manager the meta engine can post process the options string as needed before the GUI gets to see it.

The second commit puts this to good use for the SCUMM engine by adding a default render mode to the options. My aim here is supporting PR #4223 . There, the default render modes do not get added to the detection tables so as to keep the table modifications to a minimum (= the VGA games don't have a VGA option, the EGA games don't have an EGA option, the Amiga games don't have an Amiga option, etc.). Instead, with this PR here, these default render modes get added by the meta engine.

### CODE REVIEW

File: engines/scumm/detection.cpp [hunk 1]
```diff
@@ -252,6 +253,53 @@ static const ExtraGuiOption enableOriginalGUI = {
...
+	case Common::kPlatformDOS:
+		if (extra.equalsIgnoreCase("EGA") || extra.equalsIgnoreCase("V1") || extra.equalsIgnoreCase("V2"))
+			defaultRenderOption = GUIO_RENDEREGA;
+		break;
+	case Common::kPlatformUnknown:
+		// For targets that don't specify the platform we stick with default VGA.
+		break;
+	default:
+		// Clear this for platforms that don't have a specific render option (SegaCD, NES, ...)
+		defaultRenderOption = nullptr;
```
Comment (somaen) [inline engines/scumm/detection.cpp:291]: IMHO it's slightly more readable to initialize to nullptr initially, and then let the default here set VGA.

Comment (athrxx) [reply engines/scumm/detection.cpp]: Sure, I have changed it...

**Comment by eriktorbjorn** [comment]
Since this pull request is based on a more recent Git snapshot, would it make sense to bring in the relevant bits from #4223 and continue that work here?

**Comment by athrxx** [comment]
Yes, I guess it would make sense. These PRs kind of belong together...

**Comment by eriktorbjorn** [comment]
> Yes, I guess it would make sense. These PRs kind of belong together...

Will you be doing that?

**Comment by athrxx** [comment]
> > Yes, I guess it would make sense. These PRs kind of belong together...
> 
> Will you be doing that?

I now did that. But I had to resolve merge conflicts (thanks to the new GUIO_ORIGINALGUI flag). So, please check if you see anything suspicious. So far, all seems to work as intended...

**Comment by eriktorbjorn** [comment]
> I now did that. But I had to resolve merge conflicts (thanks to the new GUIO_ORIGINALGUI flag). So, please check if you see anything suspicious. So far, all seems to work as intended...

Sorry, I didn't see your reply until now. Seems to be working for me, too. (Though apparently my game entries for Maniac Mansion are old enough that I don't have the "extra" field, so for them it says VGA instead of EGA in the dropdown. I don't know if those can be automatically updated, the way we do with some other fields?)

Of course, we still have the issue that my original pull request had in that all render modes are removed for other engines. Maybe there should be a default implementation that adds VGA?

And just to re-iterate, I think the only engines that use render modes are:

- SCUMM
- Kyra
- Startrek
- AGI

So we have to make sure that those behave correctly. ("Who's this 'we' you're talking about?" :-)

**Comment by athrxx** [comment]
Okay, I just did some quick tests in this branch to understand the "issue".

It is true that all engines which don't support different render modes now only offer one list entry: "<default>". From my side, this is a behavior which I can appreciate. Users would be discouraged from making pointless selections here, since it will not change anything ingame.

However, it would require the adaption of the other engines that actually do support render modes and don't currently have GUIO flags or a custom meta engine method yet: According to your research these are AGI and Star Trek.

The reason for the behaviour which you now seem to criticize is your own change to options.cpp, line 1512. You could just get rid of that change again. Your newly added GUIO flags will still work as intended,  **you just have to run each target once**, so that the config gets updated with the new options.

**Review by sev- — APPROVED** [review]
I like this change. Speaking of AGI engine, it uses render modes for AGI games. E.g. it is a fixed set of supported platforms.

It sets up render mode based on platform by default but lets user override it with any of the 9 supported modes. See `AgiBase::initRenderMode()`.

PreAGI games are simpler: only mickey supports the manual switch to CGA (EGA is default), and troll and winnie are platform-based

**Comment by sev-** [comment]
Startrek is VGA by default with option to switch to EGA

**Comment by athrxx** [comment]
I have updated AGI with the appropriate render options. Unfortunately, I don't have any non-DOS version, but it seems to be straightforward enough....

Regarding Star Trek: I don't have that. What I would do here is just replace every case of `GUIO0()` in `startrek/detection.cpp` with `GUIO2(GUIO_RENDERVGA, GUIO_RENDEREGA)`. It would be a simple one-shot search/replace op. But as I said: I cannot test. So, am I missing something here? Or is it safe to do that?

**Comment by sev-** [comment]
It is truly safe to do that.

**Comment by athrxx** [comment]
Thanks! I have now made that change for Startrek...

**Review by bluegr — APPROVED** [review]
Nice work! This will simplify handling such options within different games

**Comment by bluegr** [comment]
Overall, great work! This has been open for almost 2 months now, and got approved.
It's in a good state to be merged. Thanks again!

**Pull request merged**